File: detectable/src/main/java/com/synopsys/integration/detectable/detectables/pip/parser/PipenvFreezeParser.java
/**
 * detectable
 *
 * Copyright (c) 2021 Synopsys, Inc.
 *
 * Use subject to the terms and conditions of the Synopsys End User Software License and Maintenance Agreement. All rights reserved worldwide.
 */
package com.synopsys.integration.detectable.detectables.pip.parser;

import java.util.List;
import java.util.stream.Collectors;

import com.synopsys.integration.detectable.detectables.pip.model.PipFreeze;
import com.synopsys.integration.detectable.detectables.pip.model.PipFreezeEntry;

public class PipenvFreezeParser {
    private static final String VERSION_SEPARATOR = "==";

    public PipFreeze parse(final List<String> pipFreezeOutput) {
        final List<PipFreezeEntry> entries = pipFreezeOutput.stream()
                                                 .map(line -> line.split(VERSION_SEPARATOR))
                                                 .filter(pieces -> pieces.length == 2)
                                                 .map(pieces -> new PipFreezeEntry(pieces[0], pieces[1]))
                                                 .collect(Collectors.toList());

        return new PipFreeze(entries);
    }
}

File: detectable/src/main/java/com/synopsys/integration/detectable/detectables/pip/parser/PipenvFreezeParser.kt
/**
 * detectable
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.detectable.detectables.pip.parser

import com.synopsys.integration.detectable.detectables.pip.model.PipFreeze
import com.synopsys.integration.detectable.detectables.pip.model.PipFreezeEntry

class PipenvFreezeParser {
    private val versionSeparator = "==".toRegex()

    fun parse(pipFreezeOutput: List<String>): PipFreeze {
        val entries = pipFreezeOutput
                .map { line -> line.split(versionSeparator) }
                .filter { pieces -> pieces.size == 2 }
                .map { pieces -> PipFreezeEntry(pieces[0], pieces[1]) }

        return PipFreeze(entries)
    }
}
